[Docs] Kibana 9.4.0 release notes (#264554)
elastic/kibana: docs/release-notes/index.md

* Improves {{esql}} editor autocomplete for full-text search functions: `MATCH_PHRASE`'s second argument now only suggests literal values, and FTS functions are excluded from `EVAL` suggestions except inside `SCORE()` [#247003]({{kib-pull}}247003).
* Improves line commenting in the {{esql}} editor to match standard IDE conventions [#254851]({{kib-pull}}254851).
* Improves query pretty printing [#257440]({{kib-pull}}257440).
* Adds an {{esql}} indentation shortcut to the editor [#247234]({{kib-pull}}247234).
* Simplifies the **Run** and **Cancel** button states in the {{esql}} editor [#254121]({{kib-pull}}254121).
* Highlights multiple word occurrences in search results [#258764]({{kib-pull}}258764).

**Elastic Observability solution**:
For the Elastic Observability 9.4.0 release information, refer to [Elastic Observability Solution Release Notes](docs-content://release-notes/elastic-observability/index.md).

**Elastic Security solution**:
For the Elastic Security 9.4.0 release information, refer to [Elastic Security Solution Release Notes](docs-content://release-notes/elastic-security/index.md).

**Kibana platform**:

> * Sets the default {{kib}} instance size to 2 GB of RAM for new {{ech}} deployments. For Platinum and Enterprise users, we recommend provisioning at least 2 GB of RAM for {{kib}} instances in all deployment types. Below this size, features such as Security Detection Rules, Reporting, Workflows, and Agent Builder might be unavailable or cause service interruptions.

Reviewer: @yuvielastic @lukeelmers Can you please have a look at this note and tell me if we need to change anything?

Reviewer 2: I would remove "might be unavailable" as features will be available always, but if customers use it then they will run into out of memory errors. 

So I will just say re-phrase this to:

Below this size, use of features such as Security Detection Rules, Reporting, Workflows, and Agent Builder  may lead to service interruptions.

Reviewer: Thanks Yuvi! Updated

Reviewer 3: This RN makes it sound like there will only be disruption if running those features at below 2GB. If running heavy workloads, its also possible to have disruptions at larger sizes. And its the use of the plat/ent features, not the license type itself. Also Reporting is disabled in ECH if <2GB (so technically its unavailable)... 

Suggest

```
Sets the default {{kib}} instance size to 2 GB of RAM for new {{ech}} deployments. This is the recommended minimum in all deployment types for Platinum and Enterprise {{kib}} features, and for production workloads. 
```

Reviewer 4: slightly reworded + made it a callout at the top of the release notes so it's more visible

* Adds a feedback button to {{kib}}'s header [#225074]({{kib-pull}}225074).
* Remembers the pagination state when navigating back from an edit on the **Users** page, instead of always returning to page 1 and resetting the search [#261152]({{kib-pull}}261152).
* Distinguishes between session idle timeouts and session lifespan timeouts [#252779]({{kib-pull}}252779).
* Improves Index Management index list load performance on large clusters with many indices [#246276]({{kib-pull}}246276).
* Adds a **Query Activity** page under Stack Management for viewing and canceling long-running queries [#253216]({{kib-pull}}253216).

**Machine Learning**:
* Updates Security ML jobs to use entity analytics fields for host and user fields [#255339]({{kib-pull}}255339).

> * Adds sparkline charts for `CATEGORIZE` in Discover [#254269]({{kib-pull}}254269).

Reviewer 4: ```suggestion
```
Removing since this is shifting to 9.5

* Adds support for unmapped fields [#248606]({{kib-pull}}248606).
* Adds support for timezone handling [#247917]({{kib-pull}}247917).

% GA commands
* Makes the `FORK` command generally available [#261904]({{kib-pull}}261904).
* Makes the `RERANK` command generally available [#252242]({{kib-pull}}252242).

% editor UI and experience
* Redesigns the {{esql}} editor interface [#251223]({{kib-pull}}251223).
* Adds a data source browser to the {{esql}} editor [#251897]({{kib-pull}}251897).
* Adds support for {{esql}} views to the editor [#261907]({{kib-pull}}261907).
* Adds {{esql}} query statistics to the editor [#251029]({{kib-pull}}251029).
* Improves {{esql}} editor autocomplete for full-text search functions: `MATCH_PHRASE`'s second argument now only suggests literal values, and FTS functions are excluded from `EVAL` suggestions except inside `SCORE()` [#247003]({{kib-pull}}247003).
* Improves line commenting in the {{esql}} editor to match standard IDE conventions [#254851]({{kib-pull}}254851).
* Improves query pretty printing [#257440]({{kib-pull}}257440).
* Adds an {{esql}} indentation shortcut to the editor [#247234]({{kib-pull}}247234).

> * Simplifies the **Run** and **Cancel** button states in the {{esql}} editor [#254121]({{kib-pull}}254121).

Reviewer 4: ```suggestion
* Simplifies the **Run** and **Cancel** button states in the {{esql}} editor [#254121]({{kib-pull}}254121).
* Highlights multiple word occurrences in search results [#258764]({{kib-pull}}258764).
```

* Makes static-lookup formatter work with aggregated boolean fields [#249311]({{kib-pull}}249311).
* Adds a check to ensure {{esql}} is valid before matching the Metrics profile [#248917]({{kib-pull}}248917).
* Prevents losing draft queries when switching tabs [#247968]({{kib-pull}}247968).
* Fixes an issue where quickly opened tabs could not complete loading [#246941]({{kib-pull}}246941).
* Fixes the default app state handling when detecting unsaved changes [#246664]({{kib-pull}}246664).

**{{esql}} editor**:
* Fixes {{esql}} multi-value filtering with `STATS` [#260998]({{kib-pull}}260998).
* Fixes `STATS` generated columns with inline `WHERE` [#260196]({{kib-pull}}260196).
* When no local indices are available, the {{esql}} query suggestion now correctly considers remote indices [#257340]({{kib-pull}}257340).
* Fixes {{esql}} variable controls not displaying server-side errors in the editor [#263020]({{kib-pull}}263020).
* Fixes autocomplete fetches piling up without cancellation when typing rapidly in the {{esql}} editor [#255664]({{kib-pull}}255664).
* Fixes incorrect validation of the `TS` (time series) command [#253635]({{kib-pull}}253635).

> * Fixes some `GROK` patterns not being recognized, which caused columns to appear as unknown [#246871]({{kib-pull}}246871).

Reviewer 4: ```suggestion
* Fixes some `GROK` patterns not being recognized, which caused columns to appear as unknown [#246871]({{kib-pull}}246871).
* Aborts in-flight long-running queries for {{esql}} controls [#254487]({{kib-pull}}254487).
* Fixes incorrect KQL bar results for some indices [#254119]({{kib-pull}}254119).
```

* Adds an auto-push case option to the case connector [#249251]({{kib-pull}}249251).
* Exposes the `maximumCasesToOpen` parameter in the case action connection [#247990]({{kib-pull}}247990).
* Adds additional workflow steps [#256922]({{kib-pull}}256922).
* Alert deletion is now generally available [#247465]({{kib-pull}}247465).

**Connectivity**:
* Adds the `region` implementation to the Bedrock Connector [#252960]({{kib-pull}}252960).
* Adds the `region` parameter to the Bedrock Connector schema [#252956]({{kib-pull}}252956).

**Dashboards and Visualizations**:
* Adds a **Badge** color option for table values in Lens, allowing cell values to be displayed as colored badges instead of text or background coloring [#257408]({{kib-pull}}257408).
* Adds a new list legend layout for horizontal legends (top and bottom), offering a more space-efficient alternative to the grid layout. This is now the default for XY charts [#257092]({{kib-pull}}257092).
* Makes the Dashboard API endpoints available as technical preview [#256302]({{kib-pull}}256302).
* Adds the ability to show and export Dashboard API JSON in a flyout [#255382]({{kib-pull}}255382).

> * Adds a new line-optimized color palette for line charts in Lens [#253437]({{kib-pull}}253437).

Reviewer 5: ```suggestion
* Adds a new "Elastic (line optimized)" color palette for line charts in Lens [#253437]({{kib-pull}}253437).
```

> * Supports {{esql}} in Vega [#247186]({{kib-pull}}247186).

Reviewer 5: ```suggestion
* Allows to use {{esql}} in Vega visualizations [#247186]({{kib-pull}}247186).
```

* Controls are now available as a panel type, allowing them to be freely placed anywhere in your dashboards [#245588]({{kib-pull}}245588).
* Adds sort order controls for heatmap visualization axes [#244696]({{kib-pull}}244696).
* Adds a middle position option for the primary metric in Metric charts in Lens [#260902]({{kib-pull}}260902).
* Improves tick labels for time-based X axes in {{esql}} heatmap visualizations [#259218]({{kib-pull}}259218).

> * Enables section dragging via the entire header and collapse/expand on click [#258502]({{kib-pull}}258502).

Reviewer 5: ```suggestion
* Enables section dragging by clicking anywhere on its header, and collapse/expand on click [#258502]({{kib-pull}}258502).
```

* Allows dragging of opened collapsible sections [#257191]({{kib-pull}}257191).
* Improves datatable visualization performance for large datasets in Lens [#256234]({{kib-pull}}256234).
* Makes the filter pills section collapsible [#255887]({{kib-pull}}255887).
* Adds a grid size gauge while resizing panels [#255363]({{kib-pull}}255363).
* Adds a borderless option to panel settings [#255021]({{kib-pull}}255021).
* Retrieves variable types from the query response [#254436]({{kib-pull}}254436).
* Defaults to line chart when the x-axis contains a timestamp, instead of a histogram [#253930]({{kib-pull}}253930).
* Suggests line charts for timeseries queries (TS / PromQL) in Lens [#252661]({{kib-pull}}252661).

> * Adds `id` as a create request parameter [#251975]({{kib-pull}}251975).

Reviewer 5: ```suggestion
* Adds `id` as an optional create request param for defining a custom ID when creating Markdown panels [#251975]({{kib-pull}}251975).
```

* Redesigns the panel titles [#251720]({{kib-pull}}251720).
* Enables monospace numeric font features for Lens renderers [#251576]({{kib-pull}}251576).
* Allows panels to be dragged while they're in focus for editing [#251327]({{kib-pull}}251327).
* Makes **Contains** the default search technique for options list controls [#250992]({{kib-pull}}250992).
* Allows IP fields to be searched using CIDR notation in controls [#250875]({{kib-pull}}250875).
* Allows filtering from legend actions when possible [#248789]({{kib-pull}}248789).
* Adds library support for the Markdown embeddable [#248779]({{kib-pull}}248779).
* Enables {{esql}} multi-terms charts in Lens [#244743]({{kib-pull}}244743).

**Data ingestion and Fleet**:

> * Allows remote ES outputs and service tokens in serverless [#262101]({{kib-pull}}262101).

Reviewer 6: ```suggestion
* Allows remote {{es}} outputs and service tokens in Serverless [#262101]({{kib-pull}}262101).
```

* Renames Cloud Connector to Federated Identity in the UX [#261353]({{kib-pull}}261353).

> * Enables OpenTelemetry (OTel) collector support in Fleet. You can now add OTel collector agents via a new **Add collector** button in the Fleet UI [#260654]({{kib-pull}}260654).

Reviewer 6: ```suggestion
* Adds support for monitoring OpenTelemetry (OTel) collectors in {{fleet}} in technical preview. You can now add OTel collector agents using the **Add** > **Collector (OpAMP)** button in the Fleet UI [#260654]({{kib-pull}}260654).
```

> * Enables the version-specific policies feature flag [#258796]({{kib-pull}}258796).

Reviewer 6: ```suggestion
* Introduces support for version-specific policies in {{fleet}} when integrations specify agent version requirements, ensuring agents receive only configurations compatible with their version [#258796]({{kib-pull}}258796).
```

> * Warns users about integrations with upcoming deprecations [#257937]({{kib-pull}}257937).

Reviewer 6: ```suggestion
* Shows UI warnings for integrations with upcoming deprecations [#257937]({{kib-pull}}257937).
```

* Resolves and merges templates listed in `template_paths` [#257730]({{kib-pull}}257730).
* Adds permission verifier background tasks [#257516]({{kib-pull}}257516).
* Installs package dependencies automatically [#256700]({{kib-pull}}256700).

> * Requests user review for auto-upgrade packages with deprecations [#255273]({{kib-pull}}255273).

Reviewer 6: ```suggestion
* Requests user review when auto-upgrading packages with deprecations [#255273]({{kib-pull}}255273).
```

* Adds an out-of-the-box alerting rule template to freshly installed integrations for monitoring idle data streams [#254730]({{kib-pull}}254730).
* Displays warnings for deprecated integration features [#253923]({{kib-pull}}253923).
* Shows warnings in the UI when an integration is deprecated [#251860]({{kib-pull}}251860).

> * Integration-managed SLO templates can now be installed by Fleet and used to create new SLOs [#250369]({{kib-pull}}250369).

Reviewer 6: ```suggestion
* Allows {{fleet}} to install integration-managed SLO templates for creating new SLOs [#250369]({{kib-pull}}250369).
```

> * Migrates inputs when the `migrate_from` field is found [#242934]({{kib-pull}}242934).

Reviewer 6: ```suggestion
* Migrates input configurations when the `migrate_from` field is specified in the package manifest [#242934]({{kib-pull}}242934).
```

> * Updates the max spec version to 3.6 [#261362]({{kib-pull}}261362).

Reviewer 6: ```suggestion
* Updates the maximum supported package specification version to 3.6 [#261362]({{kib-pull}}261362).
```

* Adds a new **Alerting** tab to the integrations UI for viewing and managing alerting-related assets [#253948]({{kib-pull}}253948).

> * Allows integration rollback if not all package policies are upgraded [#253646]({{kib-pull}}253646).

Reviewer 6: The current wording reads as if integration rollback is only possible if some integration policies haven’t been upgraded whereas the PR removes the requirement for _all_ integration policies to be upgraded for the rollback to be available.

This would be more accurate:

```suggestion
* Allows integration rollback when only some integration policies are upgraded [#253646]({{kib-pull}}253646).
```

> * Ensures time series index mode is not enabled on non-metrics time series for input packages [#251205]({{kib-pull}}251205).

Reviewer 6: ```suggestion
* Ensures the time series index mode is not enabled for input packages with non-metrics data streams [#251205]({{kib-pull}}251205).
```

> * Adds authentication fields to Elastic Agent binary download sources for connecting to self-hosted artifact registries [#250557]({{kib-pull}}250557).

Reviewer 6: ```suggestion
* Adds authentication fields to {{agent}} binary download sources managed by {{fleet}} for connecting to self-hosted artifact registries [#250557]({{kib-pull}}250557).
```

> * Improves memory usage during Fleet setup by deferring package reinstalls to async tasks [#248235]({{kib-pull}}248235).

Reviewer 6: ```suggestion
* Improves memory usage during {{fleet}} setup by deferring package reinstalls to async tasks [#248235]({{kib-pull}}248235).
```

**Kibana platform**:
* Adds the ability to configure a space-level default project routing expression in the **Space Management** UI [#250990]({{kib-pull}}250990).
* Sets the heap limit to min(60%, 4GB) for containers if unset. Containers now set the default Node.js heap to 75% of available memory up to a maximum of 4096 MB [#246073]({{kib-pull}}246073).
* Adds a feedback plugin [#225074]({{kib-pull}}225074).
* Remembers the user grid state when navigating back from edit [#261152]({{kib-pull}}261152).
* Distinguishes between session idle timeouts and session lifespan timeouts [#252779]({{kib-pull}}252779).
* Adds a new `check_package_docs` CLI for validating plugin and package API documentation. The `--stats` flag on `build_api_docs` is deprecated in favor of this new dedicated validation command [#249305]({{kib-pull}}249305).
* Adds `maxSize` for entries in security endpoints related API schemas [#246359]({{kib-pull}}246359).
* Integrates the new app menu with dashboards [#246153]({{kib-pull}}246153).

**Machine Learning**:

> * Removes the `experimental` guard from dashboards [#261530]({{kib-pull}}261530).

Reviewer 7: This one should be moved to the `Dashboards and Visualizations` section and the text edited to 

```
Enables Dashboard skill in Agent Builder
```

* Updates Security ML jobs to use entity analytics fields for host and user fields [#255339]({{kib-pull}}255339).
* Adds sparkline charts for `CATEGORIZE` in Discover [#254269]({{kib-pull}}254269).
* Adds a link to manage anomaly detection jobs in the Machine Learning left navigation [#260605]({{kib-pull}}260605).
* Anomaly detection now automatically closes the job when stopping a datafeed [#259603]({{kib-pull}}259603).

> * Adds aria labels to job wizard combo boxes [#258509]({{kib-pull}}258509).

Reviewer 7: ```suggestion
* Adds aria labels to anomaly detection job wizard combo boxes [#258509]({{kib-pull}}258509).
```

* Adds CPS-unsupported labels to ML rules [#258162]({{kib-pull}}258162).
* Adds a cross-project search warning to the anomaly detection jobs list [#256713]({{kib-pull}}256713).
* Adds CPS compatibility for AIOps and data visualizer [#256244]({{kib-pull}}256244).
* Updates the `v3_rare_process_by_host_windows` bucket span to two hours [#255855]({{kib-pull}}255855).
* Changes the rare process by host Windows job bucket span from 15m to 4h [#255385]({{kib-pull}}255385).
* Adds a new single APM Correlations endpoint for latency and failed transactions [#254607]({{kib-pull}}254607).
* Adds Gemini 2.5 Flash Lite, Claude 4.5 Haiku, and Claude 4.6 Sonnet preconfigured connectors [#253109]({{kib-pull}}253109).
* Adds a dynamic default connector in GenAI settings [#252861]({{kib-pull}}252861).

> * Replaces the custom RobotIcon (Agent) with the EUI productAgent icon in the Agent Opt-in flow [#252526]({{kib-pull}}252526).

Reviewer 7: This can be removed as it isn't a user-facing change. I switched the release note label on the PR to `release_note:skip`.

* Adds a zoom in button to the date picker [#252252]({{kib-pull}}252252).
* Adds Anthropic Claude Opus 4.6 preconfigured connector [#252177]({{kib-pull}}252177).
* Uses the `location` field to correctly set provider config in AI/Inference Connector creation [#250838]({{kib-pull}}250838).

> * Adds the `artifactRepositoryProxyUrl` setting [#250771]({{kib-pull}}250771).

Reviewer 7: ```suggestion
* Adds proxy URL setting for product documentation artifacts [#250771]({{kib-pull}}250771).
```

* Adds new preconfigured connectors [#249379]({{kib-pull}}249379).
* Moves the results view buttons closer to the job selection controls in Anomaly Detection [#249261]({{kib-pull}}249261).
* Adds missing {{esql}} commands and functions documentation for inference tasks [#249089]({{kib-pull}}249089).
* Enhances model memory estimation for supplied configurations in anomaly detection [#248479]({{kib-pull}}248479).
* Adds the `timeout` parameter to InferenceChatModel [#248326]({{kib-pull}}248326).

> * Removes usage of the AJV library and replaces it with zod [#248317]({{kib-pull}}248317).

Reviewer 7: This one can be removed.

* Consolidates Type, Preconfigured, and Tech Preview badges under the endpoint name and removes the dedicated Type column in the inference endpoints table [#252621]({{kib-pull}}252621).
* Improves AI connector setup by auto-populating the model field with recommended defaults [#250506]({{kib-pull}}250506).
* Improves the inference endpoints page by adding a Model column and enabling search by model name [#249779]({{kib-pull}}249779).
* Adds descriptions to the `semantic_text` field inference endpoint select [#249265]({{kib-pull}}249265).
* Improves the inference endpoint selector layout to keep long endpoint names readable and stable while clarifying ML-node startup behavior [#247417]({{kib-pull}}247417).
* Displays the API key tab if the user has permission, and hides it for users without API key management permissions [#246979]({{kib-pull}}246979).
* Updates the Search homepage design [#246777]({{kib-pull}}246777).

**Workflows**:
* Enables OAuth on the Jira connector [#262189]({{kib-pull}}262189).
* Enables OAuth on the Zoom connector [#260775]({{kib-pull}}260775).

> * Adds the `linux.advanced.ransomware.diagnostic` advanced flag in preparation for the Linux ransomware release [#260538]({{kib-pull}}260538).

Reviewer 8: ```suggestion
```

This should go under Security

* Enables OAuth on Notion, Figma, Sharepoint, and ServiceNow connectors [#260395]({{kib-pull}}260395).
* Enables OAuth on the Salesforce connector [#259826]({{kib-pull}}259826).
* Enables OAuth on Google connectors [#259528]({{kib-pull}}259528).

> * Creates an initialization endpoint and migrates the list index creation flow [#258891]({{kib-pull}}258891).

Reviewer 8: ```suggestion
```
Security

* Adds import and export features for workflows [#257976]({{kib-pull}}257976).
* Allows Agent Builder to detect outdated attachments and lets users refresh them into the next message [#257658]({{kib-pull}}257658).
* Adds the `workflows.executionFailed` trigger so you can run workflows when another workflow fails. Use it to send notifications (e.g., Slack), run cleanup, or trigger retries [#257633]({{kib-pull}}257633).

> * Adds leads cards, flyouts, and Agent Builder integration [#257610]({{kib-pull}}257610).

Reviewer 8: ```suggestion
```
Security

* Adds the `artifactRepositoryProxyUrl` setting [#250771]({{kib-pull}}250771).
* Adds new preconfigured connectors [#249379]({{kib-pull}}249379).
* Moves the results view buttons closer to the job selection controls in Anomaly Detection [#249261]({{kib-pull}}249261).
* Adds missing {{esql}} commands and functions documentation for inference tasks [#249089]({{kib-pull}}249089).
* Enhances model memory estimation for supplied configurations in anomaly detection [#248479]({{kib-pull}}248479).
* Adds the `timeout` parameter to InferenceChatModel [#248326]({{kib-pull}}248326).
* Removes usage of the AJV library and replaces it with zod [#248317]({{kib-pull}}248317).
* Adds time window buttons to the date picker [#248142]({{kib-pull}}248142).
* Adds a button to synchronize saved objects in trained models [#247691]({{kib-pull}}247691).
* Refreshes the Overview page [#247573]({{kib-pull}}247573).
* Marks 429 errors as user errors in Inference/AI Connector [#246640]({{kib-pull}}246640).
* Opens matching pattern docs in a new Discover tab [#245695]({{kib-pull}}245695).

> **Search**:

Reviewer 4: @leemthompo @szabosteve not sure what the best place is for all EIS/inference stuff. Search or machine learning sound fine options to me, but I think we have a mix of both right now here and in the fixes too. If you have a preference please let it be known :)

Reviewer 9: "Search AI" would probably do the trick, WDYT @florent-leborgne?

Reviewer 4: Maybe

> * Adds a serverless Trial Companion [#244178]({{kib-pull}}244178).

Reviewer 8: ```suggestion
```

Security

* Validates HTTP connector URL against `allowedHosts` at execution time [#258080]({{kib-pull}}258080).
* Uses correct URL in `viewInAppUrl` and fixes `??` bug in redirect [#257910]({{kib-pull}}257910).
* Fixes an issue where embeddables cannot load when no references are provided [#257779]({{kib-pull}}257779).
* Fixes false validation errors for item/index in `data.map` steps [#257703]({{kib-pull}}257703).
* Fixes an accessibility issue in the table with zoom [#257097]({{kib-pull}}257097).
* Adds support for native YAML arrays for foreach step [#256298]({{kib-pull}}256298).
* Fixes false Liquid validation errors from YAML comments [#256237]({{kib-pull}}256237).
* Fixes deanonymization offset drift and adds regression coverage [#256112]({{kib-pull}}256112).
* Adds a generate API key server-side implementation [#256083]({{kib-pull}}256083).
* Fixes a bug in the `platform.core.search` tool and `index_search` tool type where nested fields were ignored when searching for matching documents [#255914]({{kib-pull}}255914).
* Fixes MCP connectors ignoring the proxy/SSL configuration from the actions plugin (`xpack.actions` configuration property) [#255813]({{kib-pull}}255813).

> * Returns an error when submitting a scan job but malware protection is not enabled [#255597]({{kib-pull}}255597).

Reviewer 8: ```suggestion
```

Security

* Fixes execution tree clipping when the foreach step has many iterations [#253576]({{kib-pull}}253576).
* Fixes `data[].url.body.runtime_mappings` param being ignored or overridden in Vega specs [#253560]({{kib-pull}}253560).
* Fixes the scrollbars color in Safari when the macOS appearance setting is the opposite of the selected Kibana Color Mode [#253484]({{kib-pull}}253484).
* Fixes false validation errors for template-local variables in Liquid templates [#253405]({{kib-pull}}253405).
* Fixes Beats tutorial authentication instructions in serverless environments to show correct API key-based configuration [#253164]({{kib-pull}}253164).
* Adds datemath support to KQL evaluator [#252840]({{kib-pull}}252840).
* Fixes the create new tool page when switching types [#252811]({{kib-pull}}252811).
* Prevents table sorting when toggling the workflow enable state [#252724]({{kib-pull}}252724).
* Strips system-managed date fields from ingest pipelines before PUT [#252579]({{kib-pull}}252579).
* Fixes a UI bug with the workflow tool type [#252563]({{kib-pull}}252563).
* Propagates connector API errors [#252372]({{kib-pull}}252372).

> * Fixes Entity Store Install API parameter validations [#252366]({{kib-pull}}252366).

Reviewer 8: ```suggestion
```
Security

* Makes **Contains** the default search technique for options list controls [#250992]({{kib-pull}}250992).
* Allows IP fields to be searched using CIDR notation in controls [#250875]({{kib-pull}}250875).

% sections
* Extends the selectable area for dragging, collapsing and expanding sections to their entire header [#258502]({{kib-pull}}258502).
* Allows dragging of opened collapsible sections [#257191]({{kib-pull}}257191).

% dashboard usability
* Makes the filter pills section collapsible [#255887]({{kib-pull}}255887).
* Adds a grid size gauge while resizing panels [#255363]({{kib-pull}}255363).
* Adds a borderless option to panel settings [#255021]({{kib-pull}}255021).
* Adds library support for markdown panels [#248779]({{kib-pull}}248779).
* Allows panels to be dragged while they're in focus for editing [#251327]({{kib-pull}}251327).
* Redesigns the panel titles [#251720]({{kib-pull}}251720).
* Refreshes the Dashboards app menu [#246153]({{kib-pull}}246153).

% discover sessions in dashboards
* Adds a **Discover session** panel option to dashboards [#256293]({{kib-pull}}256293).

> * Editing an unlinked Discover session panel in a dashboard now saves changes back to that panel without affecting the original saved session [#250438]({{kib-pull}}250438).

Reviewer 10: ```suggestion
* Editing an unlinked Discover session panel in a dashboard now saves changes back to that panel [#250438]({{kib-pull}}250438).
```
The last part makes it sound like it previously overwrote the original saved session, but it just wasn't possible to edit unlinked panels. I think just dropping that should be fine.

> * Clicking on a filterable field value in a Discover {{esql}} session embedded in a dashboard now creates a DSL filter, consistent with how filtering works elsewhere [#249357]({{kib-pull}}249357).

Reviewer 10: ```suggestion
* Filtering a field value in a Discover {{esql}} session embedded in a dashboard now creates a DSL filter, consistent with how filtering works elsewhere [#249357]({{kib-pull}}249357).
```
Nit: maybe slightly more accurate since it's technically the filter button you click

* Improves memory usage during {{fleet}} setup by deferring package reinstalls to async tasks [#248235]({{kib-pull}}248235).

**Discover**:
% esql mode
* Redesigns the {{esql}} editor footer in Discover: removes the row limit and timestamp indicators, and adds query run statistics [#244284]({{kib-pull}}244284).
* Adds a fields browser to the {{esql}} editor in Discover [#252749]({{kib-pull}}252749).
* Adds a layout toggle to show or hide the data table in Discover, with the state persisted in the URL [#259083]({{kib-pull}}259083).
* Adds a grouped view in Discover for {{esql}} queries that use `STATS ... BY` with a single grouping field. A new toolbar selector lets you pivot by that field or switch back to the standard table view. [#220119]({{kib-pull}}220119).
* Converts DSL filters to {{esql}} when possible when switching to {{esql}} mode [#259260]({{kib-pull}}259260).
* Persists the query mode ({{esql}} or classic) to local storage so that the next sessions open with the last mode used [#250388]({{kib-pull}}250388).

> * Adds a feature tour to the **Tab** menu for Switch to classic / {{esql}} modes [#254183]({{kib-pull}}254183).

Reviewer 10: ```suggestion
```
We removed this shortly after merging it 😅

Reviewer 4: That explains why I couldn't find it :D and it reassures me a little

* Shows Streams field descriptions in the {{esql}} editor and field sidebar [#260582]({{kib-pull}}260582).

% metrics
* Filters from the top-level {{esql}} WHERE clause now propagate into per-metric charts in the Discover metrics grid [#249103]({{kib-pull}}249103).
* Adds support for visualizing `tdigest` and `exponential_histogram` histogram metrics in the Discover metrics grid [#249269]({{kib-pull}}249269).
* Hides the data table by default when the metrics-specific Discover experience is triggered [#260607]({{kib-pull}}260607).

% tabs and sessions
* Adds the ability to restore recently closed tab groups [#253365]({{kib-pull}}253365).
* Hovering over an entry in the recently closed tabs menu now shows a preview of what the tab contained [#246973]({{kib-pull}}246973).
* Moves the inspector menu item to the tab menu [#258767]({{kib-pull}}258767).
* The chart interval is now saved with Discover sessions and restored when reopening them [#246426]({{kib-pull}}246426).

> * Adds notifications for background search completion [#249857]({{kib-pull}}249857).

Reviewer 10: I'm not sure this one belongs here. Background searches are available in several apps including Discover and Dashboard.

Reviewer 4: Yeah, I'll try to find a better place but if nothing obvious either I'll leave it here, it's not a critical one IMO since it's going to be visible by the folks using it

* Adds a **Save Discover table to dashboard** option [#259626]({{kib-pull}}259626).

% general

> * Highlights multiple word occurrences in search results [#258764]({{kib-pull}}258764).

Reviewer 10: This is an ES|QL editor enhancement.

Reviewer 4: ```suggestion
```

> * Returns to the Discover session after **Save As** [#255323]({{kib-pull}}255323).

Reviewer 10: ```suggestion
```
Just switched this one's label to `release_note:skip`. I think labelling it an enhancement was a mistake.

> * Shows a table view for indexes and views with a small number of columns [#255292]({{kib-pull}}255292).

Reviewer 10: ```suggestion
* Adds default table columns for indexes and views with a small number of fields [#255292]({{kib-pull}}255292).
```
Maybe a bit more accurate.

% doc viewer

> * The document viewer **Table** tab now restores your previous state — including search, filters, pinned fields, and pagination — when you reopen the flyout [#249682]({{kib-pull}}249682).

Reviewer 10: ```suggestion
```
Also switched this one to `release_note:skip`. It's part of the flyout tabs feature in the next bullet.

* The doc viewer flyout now stays open when switching between Discover tabs and remembers which tab (such as **Table** or **JSON**) was active in each [#246612]({{kib-pull}}246612).

> * Uses a restorable state in doc viewer's **JSON** tab [#252054]({{kib-pull}}252054).

Reviewer 10: ```suggestion
```
Same thing here 🙃 Sorry, our team will ned to keep a closer eye on this going forward!

* Improves error handling in debug API [#258115]({{kib-pull}}258115).
* Fixes the unenroll task and adds an FTR test [#255726]({{kib-pull}}255726).
* Fixes the incorrect installation of assets [#254923]({{kib-pull}}254923).
* Filters out unenrolled agents in the cleanup policy revisions task [#254899]({{kib-pull}}254899).
* Fixes an issue where an agent rolled back after an upgrade could not be upgraded again in the Fleet UI [#253850]({{kib-pull}}253850).
* Fixes a TypeError when an integration has no SVG icons [#251308]({{kib-pull}}251308).
* Adds back support for generating a CSV report of Fleet agent data in serverless environments [#247185]({{kib-pull}}247185).

**Discover**:
* Fixes the date picker showing empty when switching from KQL to {{esql}} [#261175]({{kib-pull}}261175).
* Fixes a tab URL state leak when leaving Discover [#262929]({{kib-pull}}262929).
* Resets the time field when the updated index pattern does not have it [#262001]({{kib-pull}}262001).
* Resets the default profile state when transitioning between tab modes [#255226]({{kib-pull}}255226).

> * Makes matches expandable for long filter matches [#255093]({{kib-pull}}255093).

Reviewer 10: ```suggestion
* Makes matches cells expandable for long field filter matches [#255093]({{kib-pull}}255093).
```
Tiny suggestion, but also this is in data view management.

> * Aborts in-flight long-running queries [#254487]({{kib-pull}}254487).

Reviewer 10: ```suggestion
* Aborts in-flight long-running queries for ES|QL controls [#254487]({{kib-pull}}254487).
```
This is specific to ES|QL controls, might belong in editor section too.

Reviewer 4: ```suggestion
```

> * Fixes incorrect KQL bar results for some indices [#254119]({{kib-pull}}254119).

Reviewer 10: Also an ES|QL editor one.

Reviewer 4: ```suggestion
```

* Fixes URL, Badge, Color, and other field formatters incorrectly rendering fields with missing or null values [#251892]({{kib-pull}}251892).
* Fixes filtering out null values from the Discover histogram legend in {{esql}} mode [#249302]({{kib-pull}}249302).
* Fixes **Search entire time range** for date nanos [#248495]({{kib-pull}}248495).
* Prevents doc viewer flyout tabs from unnecessarily re-mounting on query refresh [#248203]({{kib-pull}}248203).

> * Fixes the toolbar mode selector not responding when clicking on tabs [#247836]({{kib-pull}}247836).

Reviewer 10: ```suggestion
* Fixes dropdown menus staying open when switching tabs [#247836]({{kib-pull}}247836).
```
For clarity.

* Makes static-lookup formatter work with aggregated boolean fields [#249311]({{kib-pull}}249311).
* Adds a check to ensure {{esql}} is valid before matching the Metrics profile [#248917]({{kib-pull}}248917).

> * Fixes query drafts when switching tabs [#247968]({{kib-pull}}247968).

Reviewer 10: ```suggestion
* Prevents losing draft queries when switching tabs [#247968]({{kib-pull}}247968).
```
For clarity.

* Removes usage of the AJV library and replaces it with zod [#248317]({{kib-pull}}248317).
* Adds time window buttons to the date picker [#248142]({{kib-pull}}248142).
* Adds a button to synchronize saved objects in trained models [#247691]({{kib-pull}}247691).
* Refreshes the Overview page [#247573]({{kib-pull}}247573).
* Marks 429 errors as user errors in Inference/AI Connector [#246640]({{kib-pull}}246640).
* Opens matching pattern docs in a new Discover tab [#245695]({{kib-pull}}245695).

**Management**:
* Improves Index Management index list load performance on large clusters with many indices [#246276]({{kib-pull}}246276).
* Adds an option to convert an index to a lookup index in the **Manage index** dropdown in serverless mode [#248730]({{kib-pull}}248730).

**Search**:
* Adds warnings to the **Feature Settings** page for models that are invalid [#262262]({{kib-pull}}262262).
* Deprecates search indices in favor of index management [#260210]({{kib-pull}}260210).
* Adds a **Models** page for inference management [#259374]({{kib-pull}}259374).
* Adds a **Model Settings** UI for inference endpoint assignments [#258871]({{kib-pull}}258871).

> * Adds a SharePoint Server (on-premises) connector for Workplace AI, enabling users to connect to on-prem SharePoint instances and search, browse, and retrieve documents and site content [#258014]({{kib-pull}}258014).

Reviewer 11: ```suggestion
* Adds a SharePoint Server connector, enabling users to connect to on-prem SharePoint instances and search, browse, and retrieve documents and site content [#258014]({{kib-pull}}258014).
```

* Improves the Inference Endpoints management page by adding a view to group by models, making this the default view [#252984]({{kib-pull}}252984).
* Consolidates Type, Preconfigured, and Tech Preview badges under the endpoint name and removes the dedicated Type column in the inference endpoints table [#252621]({{kib-pull}}252621).
* Improves AI connector setup by auto-populating the model field with recommended defaults [#250506]({{kib-pull}}250506).
* Improves the inference endpoints page by adding a Model column and enabling search by model name [#249779]({{kib-pull}}249779).
* Adds descriptions to the `semantic_text` field inference endpoint select [#249265]({{kib-pull}}249265).
* Improves the inference endpoint selector layout to keep long endpoint names readable and stable while clarifying ML-node startup behavior [#247417]({{kib-pull}}247417).
* Displays the API key tab if the user has permission, and hides it for users without API key management permissions [#246979]({{kib-pull}}246979).
* Updates the Search homepage design [#246777]({{kib-pull}}246777).

**Workflows**:
* Enables OAuth on the Jira connector [#262189]({{kib-pull}}262189).

> * Enables OAuth on the Zoom connector [#260775]({{kib-pull}}260775).

Reviewer 11: I think we probably want to focus on the fact that we _added_ these connectors (they didn't exist before) and less about what features are present?

Reviewer 4: 💯

* Adds the `linux.advanced.ransomware.diagnostic` advanced flag in preparation for the Linux ransomware release [#260538]({{kib-pull}}260538).
* Enables OAuth on Notion, Figma, Sharepoint, and ServiceNow connectors [#260395]({{kib-pull}}260395).
* Enables OAuth on the Salesforce connector [#259826]({{kib-pull}}259826).

> * Enables OAuth on Google connectors [#259528]({{kib-pull}}259528).

Reviewer 11: Same

* Creates an initialization endpoint and migrates the list index creation flow [#258891]({{kib-pull}}258891).
* Adds import and export features for workflows [#257976]({{kib-pull}}257976).
* Adds the `workflows.executionFailed` trigger so you can run workflows when another workflow fails. Use it to send notifications (e.g., Slack), run cleanup, or trigger retries [#257633]({{kib-pull}}257633).
* Adds leads cards, flyouts, and Agent Builder integration [#257610]({{kib-pull}}257610).

> * Adds Confluence Cloud as a data source for Workplace AI. Users can connect to Confluence Cloud (REST API v2) with Basic auth and use agent tools to list and retrieve spaces and pages [#256508]({{kib-pull}}256508).

Reviewer 11: ```suggestion
* Adds Confluence Cloud connector. Users can connect to Confluence Cloud and use agent tools to list and retrieve spaces and pages [#256508]({{kib-pull}}256508).
```

> * Adds Confluence Cloud as a data source for Workplace AI. Users can connect to Confluence Cloud (REST API v2) with Basic auth and use agent tools to list and retrieve spaces and pages [#256508]({{kib-pull}}256508).

Reviewer 11: I think the Sharepoint Server connector line from the Search section probably belongs down here with these

* Adds a Figma data source [#255322]({{kib-pull}}255322).
* Adds a Zoom connector and data source [#255174]({{kib-pull}}255174).
* Adds a PagerDuty MCP implementation [#255154]({{kib-pull}}255154).
* Adds a Firecrawl data source [#255004]({{kib-pull}}255004).
* Adds a Zendesk data source [#254739]({{kib-pull}}254739).
* Adds a server-side workflow validation endpoint [#254502]({{kib-pull}}254502).

> * Adds a connector for Amazon S3 list and get file from bucket [#253753]({{kib-pull}}253753).

Reviewer 11: We seem to be missing quite a few. CC @talboren , since some of these are yours, not sure if they were intentionally ommitted.

```
  1. abuseipdb                                                                                                                                                        
  2. alienvault_otx                                             
  3. aws_lambda                                                                                                                                                     
  4. azure_blob                                                                                                                                                     
  5. brave_search
  6. gcp_cloud_functions
  7. github                                                                                                                                                           
  8. gmail
  9. google_cloud_storage                                                                                                                                             
  10. google_drive                                              
  11. greynoise                          
  12. jina                                                                                                                                                          
  13. microsoft_teams
  14. one_password
  15. shodan
  16. urlvoid                                                                                                                                                         
  17. virustotal
```

Reviewer 4: @seanstory I was just reviewing that and was thinking the same. PRs need 2 mandatory things and 1 optional thing to be correctly interpreted and placed in the release notes without manual intervention:
1. A proper release_note label other than "skip" (for those I assume "feature" is the right one)
2. A team or feature label (we use that to automatically map to categories/sections in the release notes)
3. (optional) A `## Release notes` section in the PR description that says what the feature is/adds for users -> This is where you can avoid things like "Workplace AI" and elaborate slightly more than in PR titles. Otherwise the tool falls back to the PR title (which is fine if it's clean :D)

Reviewer 4: ^^ That may be easy to automate since you're already relying on skills to populate the docs in those PRs - I actually didn't pay attention to that when reviewing them 😅

Reviewer 11: Yeah, I expect that folks have built a very bad habit of doing `release_note:skip` :( The agent skills my team has been using on these dont' go as far as creating the PRs themselves, so I'm afraid I can only blame this on user-error, and therefore means I can't easily automate it away. I'll raise it with my team going forward, though.
Do you need me to track down all the PRs in question? Or can we just consolidate all these into a single line in the release notes like:
```
- Connectors added: abuseIpDB, AlienVault OTX, AWS Lambda ....
```
?

Reviewer 4: I should be able to track them; Would you prefer that they stay under Workflows or move to Connectivity? (I'm moving them but I'll revert if needed depending on your preference)

Reviewer 4: I'll also drop the Oauth entries, since the connectors are new anyways, we don't need extra entries to say they got improved before they officially "release" in 9.4

Reviewer 4: Made some updates and moved those to the Connectivity section - the entries are simple and are making the repetitiveness look a bit ugly but lmk what you think. For those that I could not easily track down an implementation PR, I used the PR that adds their docs as fallback, to stay consistent with the pattern of having a PR to link to.

Reviewer 11: Works for me! Thanks!

* Improves {{esql}} editor autocomplete for full-text search functions: `MATCH_PHRASE`'s second argument now only suggests literal values, and FTS functions are excluded from `EVAL` suggestions except inside `SCORE()` [#247003]({{kib-pull}}247003).
* Improves line commenting in the {{esql}} editor to match standard IDE conventions [#254851]({{kib-pull}}254851).
* Improves query pretty printing [#257440]({{kib-pull}}257440).
* Adds an {{esql}} indentation shortcut to the editor [#247234]({{kib-pull}}247234).
* Simplifies the **Run** and **Cancel** button states in the {{esql}} editor [#254121]({{kib-pull}}254121).
* Highlights multiple word occurrences in search results [#258764]({{kib-pull}}258764).

**Elastic Observability solution**:
For the Elastic Observability 9.4.0 release information, refer to [Elastic Observability Solution Release Notes](docs-content://release-notes/elastic-observability/index.md).

**Elastic Security solution**:
For the Elastic Security 9.4.0 release information, refer to [Elastic Security Solution Release Notes](docs-content://release-notes/elastic-security/index.md).

**Kibana platform**:

> * Sets the default {{kib}} instance size to 2 GB of RAM for new {{ech}} deployments. For Platinum and Enterprise users, we recommend provisioning at least 2 GB of RAM for {{kib}} instances in all deployment types. Below this size, features such as Security Detection Rules, Reporting, Workflows, and Agent Builder might cause service interruptions.

Reviewer 12: can we move this up somewhere more prominent? If we don't label it as breaking, we should at least try to put it toward the top of the release notes. I worry it's too buried if we just stick it in the platform features section. there are users who will need to take action before upgrading to ensure they continue to have the best experience possible.

also optional: could be more explicit `in all deployment types beginning in 9.4.0`

Reviewer 4: Moved it higher, just under the 9.4 section title, and made it a callout
